llvm/include/llvm/Analysis/KernelInfo.h

```diff
@@ -0,0 +1,121 @@
+//=- KernelInfo.h - Kernel Analysis -------------------------------*- C++ -*-=//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the KernelInfo, KernelInfoAnalysis, and KernelInfoPrinter
+// classes used to extract function properties from a GPU kernel.
+//
+// See llvm/docs/KernelInfo.rst.
+// ===---------------------------------------------------------------------===//
+
+#ifndef LLVM_ANALYSIS_KERNELINFO_H
+#define LLVM_ANALYSIS_KERNELINFO_H
+
+#include "llvm/Analysis/OptimizationRemarkEmitter.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
+
+namespace llvm {
+class DominatorTree;
+class Function;
+
+/// Data structure holding function info for kernels.
+class KernelInfo {
+  void updateForBB(const BasicBlock &BB, int64_t Direction,
+                   OptimizationRemarkEmitter &ORE);
+
+public:
+  static KernelInfo getKernelInfo(Function &F, FunctionAnalysisManager &FAM);
+
+  bool operator==(const KernelInfo &FPI) const {
+    return std::memcmp(this, &FPI, sizeof(KernelInfo)) == 0;
+  }
+
+  bool operator!=(const KernelInfo &FPI) const { return !(*this == FPI); }
+
+  /// Whether the function has external linkage and is not a kernel function.
+  bool ExternalNotKernel = false;
+
+  /// OpenMP Launch bounds.
+  ///@{
+  std::optional<int64_t> OmpTargetNumTeams;
+  std::optional<int64_t> OmpTargetThreadLimit;
+  ///@}
+
+  /// AMDGPU launch bounds.
+  ///@{
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsX;
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsY;
+  std::optional<int64_t> AmdgpuMaxNumWorkgroupsZ;
+  std::optional<int64_t> AmdgpuFlatWorkGroupSizeMin;
+  std::optional<int64_t> AmdgpuFlatWorkGroupSizeMax;
+  std::optional<int64_t> AmdgpuWavesPerEuMin;
+  std::optional<int64_t> AmdgpuWavesPerEuMax;
```

Comment:
Instead of optional, just initialize to the maximum values?

Reply:
How is that better?

Reply:
Because these aren't optional and you avoid the noise of using optional. There is always a definitive range of values for these

Reply:
> Because these aren't optional

How do you mean?  Sometimes they do not appear in the IR.

Reply:
Just because there isn't an explicit attribute doesn't mean the value doesn't exist (actually the attributes can be used to specify impossible values, which then need to be logically clamped). The logical values always exist.

For reference, AMDGPUAttributor queries the logical values for these when doing the propagation: https://github.com/llvm/llvm-project/blob/f668708796b981733a5816f2efed0d5195af923d/llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp#L186

Reply:
OK, so your argument isn't about std::optional vs. max value.  It's about reporting just what's explicit in the IR vs. the effective value.  Right?  @jdoerfert : What did you want to see here?

Reply:
We should likely query the "default/logical" values when there is no explicit one. If we run after AMDGPUAttributor, we could ensure (or maybe it is already ensured) that we make all values explicit.

Reply:
After merging a more recent main, I see KernelInfoPrinter run after AMDGPUAttributorPass when doing this:

```
$ clang -O3 -g -fopenmp --offload-arch=native test.c -foffload-lto \
    -Rpass=kernel-info -mllvm -kernel-info-end-lto \
    -Xoffload-linker --lto-debug-pass-manager
```

Only AmdgpuFlatWorkGroupSizeMin and AmdgpuFlatWorkGroupSizeMax are set.  Should I be seeing more?

Reply:
I tried a real application (OpenMC) to get a larger sample of functions.  I found that only amdgpu-flat-work-group-size was set on some functions, only amdgpu-waves-per-eu was set on some functions, both were set on some functions, and amdgpu-max-num-workgroups was not set on any function.

Is this expected behavior for now?  If so, should AMDGPUAttributor be changed, or should I extend kernel-info to query the effective values?

Reply:
I don't understand the context of these statements. Where did you see these? I don't what changed AMDGPUAttributor would need to make. amdgpu-flat-work-group-size and amdgpu-waves-per-eu are commonly set by users to some value.  amdgpu-max-num-workgroups  isn't really implemented.

After AMDGPUAttributor runs, the effective values will be propagated. What is the expected behavior if this pass runs before that? Is that what you were looking at?

Reply:
As I mentioned above, I'm running KernelInfoPrinter after AMDGPUAttributorPass.

@jdoerfert said:

> We should likely query the "default/logical" values when there is no explicit one. If we run after AMDGPUAttributor, we could ensure (or maybe it is already ensured) that we make all values explicit.

Perhaps I misunderstood what the various occurrences of "we" mean.  I read it as, if AMDGPUAttributor were to make all values explicit, then KernelInfoPrinter wouldn't have to query the "default/logical" values.

Reply:
Given that AMDGPUAttributor doesn't appear to always make these values explicit, I added [a commit](https://github.com/llvm/llvm-project/pull/102944/commits/62d494d9a9f13e5b58d71f083a6cb9f67f19579b) to query the values.

Reply:
I think this issue is handled even though the implementation details are still under discussion elsewhere.  Resolving.  Please reopen if necessary.